File: 2019/day01.hs
#!/usr/bin/env runhaskell

main :: IO ()
main = do
    input <- readFile "input/day01.txt"
    print (part1 input)
    print (part2 input)

part1 input = (sum (map (fuelRequired . read) (lines input)))
part2 input = (sum (map (recursiveFuelRequired . read) (lines input)))

fuelRequired mass = mass `div` 3 - 2

recursiveFuelRequired mass 
    | fuelRequired mass <= 0 = 0
    | otherwise = (fuelRequired mass) + recursiveFuelRequired (fuelRequired mass)

File: 2022/day1/src/main/kotlin/Day01.kt
package day01

fun part1(input: String): Long {
    val elfCaloriesCarried: List<List<String>> = input.split("\n\n").map { it.lines() }
    val elfTotals = elfCaloriesCarried.map { elf -> elf.sumOf { it.toLong() } }
    return elfTotals.max()
}

fun part2(input: String): Long {
    val elfCaloriesCarried: List<List<String>> = input.split("\n\n").map { it.lines() }
    val elfTotals = elfCaloriesCarried.map { elf -> elf.sumOf { it.toLong() } }
    return elfTotals.sortedDescending().take(3).sum()
}